database/sqlite/sqlite_metadata.c

```diff
@@ -0,0 +1,1458 @@
+}
+
+void metaqueue_store_host_labels(const char *machine_guid)
+{
+    const DICTIONARY_ITEM *acquired_host = dictionary_get_and_acquire_item(rrdhost_root_index, machine_guid);
+    queue_metadata_cmd(METADATA_STORE_HOST_LABELS, acquired_host, NULL);
+}
+
+void metaqueue_buffer(BUFFER *buffer)
+{
+    queue_metadata_cmd(METADATA_STORE_BUFFER, buffer, NULL);
+}
+
+void metaqueue_chart_labels(RRDSET *st)
+{
+    const DICTIONARY_ITEM *acquired_st = dictionary_get_and_acquire_item(st->rrdhost->rrdset_root_index, string2str(st->id));
+    queue_metadata_cmd(METADATA_ADD_CHART_LABEL, acquired_st, NULL);
+}
+
+
+//
+// unitests
+//
+
+struct thread_unittest {
+    int join;
+    unsigned added;
+    unsigned processed;
+    unsigned *done;
+};
+
+static void *unittest_queue_metadata(void *arg) {
+    struct thread_unittest *tu = arg;
+
+    struct metadata_cmd cmd;
+    cmd.opcode = METADATA_UNITTEST;
+    cmd.param[0] = tu;
+    cmd.param[1] = NULL;
+    cmd.completion = NULL;
+    metadata_enq_cmd(&metasync_worker, &cmd);
+
+    do {
+        __atomic_fetch_add(&tu->added, 1, __ATOMIC_SEQ_CST);
+        if (metadata_database_enq_cmd_noblock(&metasync_worker, &cmd))
+            break;
+        usleep(10000);
+    } while (!__atomic_load_n(&tu->join, __ATOMIC_RELAXED));
+    return arg;
+}
+
+static void *metadata_unittest_threads(int items __maybe_unused)
+{
+
+    unsigned done;
+
+    struct thread_unittest tu = {
+        .join = 0,
+        .added = 0,
+        .processed = 0,
+        .done = &done,
+    };
+
+    // Queue messages / Time it
+    time_t seconds_to_run = 5;
+    int threads_to_create = 4;
+    fprintf(
+        stderr,
+        "\nChecking metadata queue using %d threads for %ld seconds...\n",
+        threads_to_create,
+        seconds_to_run);
+
+    netdata_thread_t threads[threads_to_create];
+    tu.join = 0;
+    for (int i = 0; i < threads_to_create; i++) {
+        char buf[100 + 1];
+        snprintf(buf, 100, "meta%d", i);
+        netdata_thread_create(
+            &threads[i],
+            buf,
+            NETDATA_THREAD_OPTION_DONT_LOG | NETDATA_THREAD_OPTION_JOINABLE,
+            unittest_queue_metadata,
+            &tu);
+    }
+    uv_async_send(&metasync_worker.async);
+    sleep_usec(seconds_to_run * USEC_PER_SEC);
+
+    __atomic_store_n(&tu.join, 1, __ATOMIC_RELAXED);
+    for (int i = 0; i < threads_to_create; i++) {
+        void *retval;
+        netdata_thread_join(threads[i], &retval);
+    }
+    fprintf(stderr, "Added %u elements, processed %u\n", tu.added, tu.processed);
+
+    uv_async_send(&metasync_worker.async);
+
+    sleep_usec(5 * USEC_PER_SEC);
+
+    fprintf(stderr, "Added %u elements, processed %u\n", tu.added, tu.processed);
+
+    return 0;
+}
+
+int metadata_unittest(int items)
+{
+    metadata_sync_init();
+
+    // Queue items
+    metadata_unittest_threads(items);
+
+    fprintf(stderr, "Items still in queue %u\n", metasync_worker.queue_size);
+    metadata_sync_shutdown();
+
```

Comment:
I believe this is a test in the sense that if there's some error/bug, some assertion within the code will kill the program. I've ported most of the existing unit tests to google's test framework, and for that I need to have explicit conditions that make a test succeed/fail. I believe the intention here is to ensure the queue size is 0.

Reply:
yes the way it is not is to check if after a shutdown the queue size is 0.